### PR TITLE
Input

### PR DESCRIPTION
**Pull request by PurpurTentakel97**
implement mouse and key input.

### CODE REVIEW

File: src/silly_sprites/opengl/input.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		// single frame
+		[[nodiscard]] static bool is_key_down(Keys key);
+		[[nodiscard]] static bool is_key_up(Keys key);
+
+		// multi-frame
+		[[nodiscard]] static bool is_key_pressed(Keys key);
+		[[nodiscard]] static bool is_key_released(Keys key);
+
+		[[nodiscard]] static bool is_key_hold(Keys key);
+		[[nodiscard]] static bool is_key_raised(Keys key);
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.h:45]: What does "single frame" and "multi-frame" mean here? I would expect that `is_key_down` returns `true` if the key is currently being pressed (either because it has been pressed this frame, or if it has been pressed before and is still held down). On the other hand, I'd expect `is_key_pressed` to only return `true` if the key has been pressed down during the current frame.
Why are there six different functions instead of only four? 🤔 `...hold` and `...raised` confuse me here.

File: src/silly_sprites/opengl/input.cpp
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:1]: Please review the formatting of this file. There are newlines inside functions, but no newlines between different functions (and other small problems).

File: src/silly_sprites/opengl/input.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    void Input::update_keys(Window const& window) {
+    
+        std::swap(current_keys, last_keys);
+        current_keys.fill(false);
+    
+        for (int i = 0; i < current_keys.size(); ++i) {
+            auto const key = static_cast<Keys>(i);
+            auto const g_key = glfwGetKey(window.get(), key_to_glfw(key));
+            if (g_key == GLFW_PRESS) { current_keys[i] = true; }
+            }
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:13]: wrong indentation

File: src/silly_sprites/opengl/input.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
+#include "input.h"
+
+namespace sly::gl {
+    void Input::update_keys(Window const& window) {
+    
+        std::swap(current_keys, last_keys);
+        current_keys.fill(false);
+    
+        for (int i = 0; i < current_keys.size(); ++i) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:9]: Should be `std::size_t` instead of `int` (or `usize`, according to the styleguide, but there is no header with those type definitions yet, I will open an issue for that).

File: src/silly_sprites/opengl/input.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            auto const key = static_cast<Keys>(i);
+            auto const g_key = glfwGetKey(window.get(), key_to_glfw(key));
+            if (g_key == GLFW_PRESS) { current_keys[i] = true; }
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:12]:
```suggestion
            if (g_key == GLFW_PRESS) {
                current_keys[i] = true;
            }
```

File: src/silly_sprites/opengl/input.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            }
+    }
+    void Input::update_mouse(Window const& window) {
+        std::swap(current_mouse, last_mouse);
+        current_mouse.fill(false);
+
+        for (int i = 0; i < current_mouse.size(); ++i) { 
+            auto const key = static_cast<Mouse>(i);
+            auto const g_key = glfwGetMouseButton(window.get(), mouse_to_glfw(key));
+            if (g_key == GLFW_PRESS) { current_mouse[i] = true; }
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:22]:
```suggestion
            if (g_key == GLFW_PRESS) {
                current_mouse[i] = true;
            }
```

File: src/silly_sprites/opengl/input.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            auto const key = static_cast<Keys>(i);
+            auto const g_key = glfwGetKey(window.get(), key_to_glfw(key));
+            if (g_key == GLFW_PRESS) { current_keys[i] = true; }
+            }
+    }
+    void Input::update_mouse(Window const& window) {
+        std::swap(current_mouse, last_mouse);
+        current_mouse.fill(false);
+
+        for (int i = 0; i < current_mouse.size(); ++i) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:19]: `std::size_t` instead of `int`

File: src/silly_sprites/opengl/input.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        for (int i = 0; i < current_mouse.size(); ++i) { 
+            auto const key = static_cast<Mouse>(i);
+            auto const g_key = glfwGetMouseButton(window.get(), mouse_to_glfw(key));
+            if (g_key == GLFW_PRESS) { current_mouse[i] = true; }
+        }
+    }
+    void Input::update_mouse_position(Window const& window) {
+        glfwGetCursorPos(window.get(), &mouse_position.first, &mouse_position.second);
+    }
+    [[nodiscard]] bool Input::is_quit() {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:28]: as said before, this should be removed

File: src/silly_sprites/opengl/input.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            if (close) { return true; }
+        }
+        return false;
+    }
+
+    void Input::update(Window const& window) {
+        update_keys(window);
+        update_mouse(window);
+        update_mouse_position(window);
+        if (is_quit()) { glfwSetWindowShouldClose(window.get(), true); }
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:43]: should be removed

File: src/silly_sprites/opengl/input.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		static inline std::array<bool, magic_enum::enum_count<Mouse>()> last_mouse{  };
+
+		static inline std::vector<std::vector<Keys>> close_input {{ 
+			{ Keys::KeyLeftAlt, Keys::KeyF4 },
+			{ Keys::KeyLeftControl, Keys::KeyLeftAlt, Keys::KeyF4 },
+			{ Keys::KeyLeftShift, Keys::KeyLeftAlt, Keys::KeyF4 },
+		}};
+
+		using mouse_pos_ty = std::pair<double, double>;
+		static inline mouse_pos_ty mouse_position{ };
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.h:24]:
```suggestion
		static inline mouse_pos_ty s_mouse_position{ }; 
```

File: src/silly_sprites/opengl/input.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    [[nodiscard]] Input::mouse_pos_ty Input::get_mouse_position() {
+        return mouse_position;
+    }
+
+    [[nodiscard]] bool Input::is_key_down(Keys key) {
+        return current_keys[static_cast<int>(key)];
+    }
+    [[nodiscard]] bool Input::is_key_up(Keys key) {
+        return not current_keys[static_cast<int>(key)];;
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:54]:
```suggestion
        return not current_keys[static_cast<int>(key)];
```

File: src/silly_sprites/opengl/input.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    [[nodiscard]] bool Input::is_key_pressed(Keys key) {
+            return not last_keys[static_cast<int>(key)] and current_keys[static_cast<int>(key)];
+    }
+    [[nodiscard]] bool Input::is_key_released(Keys key) {
+            return last_keys[static_cast<int>(key)] and not current_keys[static_cast<int>(key)];
+    }
+
+    [[nodiscard]] bool Input::is_key_hold(Keys key) {
+            return last_keys[static_cast<int>(key)] and current_keys[static_cast<int>(key)];
+    }
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:66]: Now I see what this function does. But why should we need this? I think this (and `is_key_raised`) is just confusing.

File: src/silly_sprites/opengl/mouse.hpp [hunk 12]
```diff
@@ -0,0 +1,62 @@
+#pragma once
+namespace sly::gl {
+    enum class Mouse {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/mouse.hpp:3]: This enum should be renamed to `MouseButton`, I think.

File: src/silly_sprites/opengl/input.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    [[nodiscard]] bool Input::is_mouse_hold(Mouse mouse) {
+            return last_mouse[static_cast<int>(mouse)] and current_mouse[static_cast<int>(mouse)];
+    }
+    [[nodiscard]] bool Input::is_mouse_raised(Mouse mouse) {
+        return not last_mouse[static_cast<int>(mouse)] and not current_mouse[static_cast<int>(mouse)];
+    }
+
+    [[nodiscard]] bool Input::add_close_input(std::vector<Keys> const& keys) {
+
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:93]: unnecessary newline

File: src/silly_sprites/opengl/input.cpp [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    [[nodiscard]] bool Input::add_close_input(std::vector<Keys> const& keys) {
+        
+        auto const contains{[keys](Keys key) {
+            for (auto const& k : keys) {
+                if (k == key) { return true; }
+            }
+            return false;
+        }};
```
Comment (mgerhold) [inline src/silly_sprites/opengl/input.cpp:99]: The styleguide required the use of `=`. Also, this lambda captures the vector by value (so it takes an expensive copy). It should capture the vector by reference. You can also utilize the `algorithm` header to simplify this.
```suggestion
        auto const contains = [&keys](Keys key) {
            return std::find(std::begin(keys), std::end(keys), key) != std::end(keys);
        };
```

File: src/silly_sprites/opengl/keys.hpp [hunk 15]
```diff
@@ -0,0 +1,619 @@
...
+        KeyLeftControl,
+        KeyLeftAlt,
+        KeyLeftSuper,
+        KeyRightShift,
+        KeyRightControl,
+        KeyRightAlt,
+        KeyRightSuper,
+        KeyMenu,
+    };
+    [[nodiscard]] constexpr Keys glfw_to_key(int glfw) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/keys.hpp:125]:
```suggestion
    [[nodiscard]] constexpr Keys glfw_to_key(int const glfw) {
```
(must be changed in the script)

Comment (mgerhold) [inline src/silly_sprites/opengl/keys.hpp:125]: There should be a newline between those.

File: src/silly_sprites/opengl/keys.hpp [hunk 16]
```diff
@@ -0,0 +1,619 @@
...
+            case GLFW_KEY_RIGHT_SUPER:
+                return Keys::KeyRightSuper;
+            case GLFW_KEY_MENU:
+                return Keys::KeyMenu;
+            default:
+                assert(false);
+                throw std::runtime_error("invalid glfw key define");
+        }
+    }
+    [[nodiscard]] constexpr int key_to_glfw(Keys key) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/keys.hpp:372]: I'd also add a newline here.

File: src/silly_sprites/opengl/mouse.hpp [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        Key4,
+        Key5,
+        Key6,
+        Key7,
+        Key8,
+        KeyLeft = Key1,
+        KeyRight = Key2,
+        KeyMiddle = Key3,
+    };
+    [[nodiscard]] constexpr Mouse glfw_to_mouse(int glfw) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/mouse.hpp:16]: newline

File: src/silly_sprites/opengl/mouse.hpp [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+            case GLFW_MOUSE_BUTTON_7:
+                return Mouse::Key7;
+            case GLFW_MOUSE_BUTTON_8:
+                return Mouse::Key8;
+            default:
+                assert(false);
+                throw std::runtime_error("invalid glfw mouse define");
+        }
+    }
+    [[nodiscard]] constexpr int mouse_to_glfw(Mouse key) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/mouse.hpp:39]: newline

File: src/silly_sprites/opengl/keys.hpp (hunk 16 above)
Comment (mgerhold) [inline src/silly_sprites/opengl/keys.hpp:372]:
```suggestion
    [[nodiscard]] constexpr int key_to_glfw(Keys const key) {
```

**Review by mgerhold — CHANGES_REQUESTED** [review]

File: src/silly_sprites/opengl/input.h (hunk 1 above)
Comment (kgnet88) [reply src/silly_sprites/opengl/input.h:45]: hey we only needed like 2h discussion to get to a version everyone could live with... 😄

**single frame:** action happens in exactly one frame so current key is either pressed or not

> down -> current
> up -> not current

**multiple frames:** action happens over two frames so 

> pressed -> not last && current
> released -> last && not current
> hold -> last && current
> raised -> not last && not current

do not give the poor purple a hard time for it, chat was on fire🔥

Comment (mgerhold) [reply src/silly_sprites/opengl/input.h:45]: He already told me that the chat was rather "demanding" today 🙈

But back to topic: What would be the use-case where you want to know if a given key is pressed **and** it was pressed during the last frame? When creating a game, there are just two cases of interest:
* was the key pressed down during this frame? (=> trigger some action)
* is the key being pressed (or held) down right now? (=> do some continuous stuff)
(same applies for released keys, of course)

I don't see a use-case where I would want to know "is the key being pressed (or held) down right now **and** was it being pressed during the last frame, too". Even if there was a use-case for that, a user could easily implement that without direct API support.

tl;dr: In my opinion, `hold` and `raised` are not needed and only lead to confusion.

Comment (kgnet88) [reply src/silly_sprites/opengl/input.h:45]: In short: that was purples stand to begin, until we confused him witch chat discussions 😄  

For what it is worth, I support you fully (given that was also my stand😈) and thankfully the ultimate power lies with you two 🙇‍♂️

File: src/scrips/generate_keys.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+path_ = "../silly_sprites/opengl/keys.hpp"
+
+# Write the code to a C++ header file
+with open(path_, "w") as file:
+    file.write("#pragma once\n namespace sly::gl {" + enum_code + switch_glfw_key + switch_key_glfw + '}\n')
+
+print("Custom enum and switch case code for GLFW keys generated successfully!")
+
+lc = ["clang-format","-i", path_] 
+retcode=call(lc)
```
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/scrips/generate_keys.py:40]: its up to you to remove this when it is implemented in CMake.

File: src/scrips/glfw3.h
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/scrips/glfw3.h:1]: Delete the header file please when cmake updates the header automatically.
or when we want to use the build directory for the scrips cmake can call that too.

File: src/silly_sprites/opengl/input.cpp (hunk 3 above)
Comment (PurpurTentakel97) [reply src/silly_sprites/opengl/input.cpp:9]: we need to keep this in mind an change it when the header exists.

File: src/silly_sprites/opengl/keys.hpp (hunk 16 above)
Comment (GibMirRechte) [reply src/silly_sprites/opengl/keys.hpp:372]: Senf :)

File: src/scrips/generate_keys.py [hunk 20]
```diff
@@ -0,0 +1,44 @@
+import re
+import sys
+from subprocess import call
+
+# Fetch the GLFW key code header from the GLFW repository:
+# https://github.com/glfw/glfw/blob/master/include/GLFW/glfw3.h
```
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/scrips/generate_keys.py:6]: cmake should update the header or set the input path to the build directory.
but than you must build a project to be able to run the python scrips.

File: tools/generate_mouse.py [hunk 21]
```diff
@@ -0,0 +1,53 @@
+import re
+import sys
+from subprocess import call
+
+# Fetch the GLFW key code header from the GLFW repository:
+# https://github.com/glfw/glfw/blob/master/include/GLFW/glfw3.h
+if len(sys.argv) < 3:
+    sys.exit(1)
+_,in_file, out_file = sys.argv
```
Comment (mgerhold) [inline tools/generate_mouse.py:9]:
```suggestion
_, in_file, out_file = sys.argv
```

File: tools/generate_mouse.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
+import re
+import sys
+from subprocess import call
+
+# Fetch the GLFW key code header from the GLFW repository:
+# https://github.com/glfw/glfw/blob/master/include/GLFW/glfw3.h
+if len(sys.argv) < 3:
```
Comment (mgerhold) [inline tools/generate_mouse.py:7]:
```suggestion
if len(sys.argv) != 3:
```

**Review by mgerhold — CHANGES_REQUESTED** [review]

File: src/silly_sprites/opengl/input.hpp [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        static constexpr auto num_key = magic_enum::enum_count<Key>();
+        static constexpr auto num_mouse = magic_enum::enum_count<MouseButton>();
+
+        static inline std::array<bool, num_key> s_current_keys{};
+        static inline std::array<bool, num_key> s_last_keys{};
+
+        static inline std::array<bool, num_mouse> s_current_mouse{};
+        static inline std::array<bool, num_mouse> s_last_mouse{};
+
+        using Vec2 = glm::vec2;
```
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/silly_sprites/opengl/input.hpp:21]: why didnt i change that? LUL

**Review by mgerhold — DISMISSED** [review]

File: src/silly_sprites/opengl/input.hpp [hunk 24]
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include "../types.hpp"
```
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/silly_sprites/opengl/input.hpp:2]: tryed it bevore but it does not work.
`D:\dev\cpp\silly-sprites\src\silly_sprites\opengl\input.hpp(2): fatal error C1083: Cannot open include file: 'types.hpp': No such file or directory`

**Review by mgerhold — APPROVED** [review]
